[DEV 2755] Client connection thread safety

morestatic opened the pull request:
This large PR introduces a number of changes that significantly improve `rportd` startup times and reliability with large numbers of clients (> 5K).

The original youtrack ticket is here but doesn't provide much further info.
https://tracker.rport.io/issue/DEV-2755/client-connection-thread-safety

### Background

`rportd` was having a number of serious performance and reliability issues when handling large numbers of clients (>1K). Particularly, a potential customer with 5K clients was finding that `rportd` wouldn't start reliably. Additionally, there have been customers reporting client disconnect problems. Also, when trying to deploy debug builds to the `node1` server, `rportd` was struggling to start due to the additional performance requirements of debug builds with race detection (typically debug builds require from x2 to x20 more performance capacity). When `rportd` did start, there were clearly data race issues. We were also seeing queues of writes to the sqlite clients.db.

Ultimately the goal of the task was to improve the rportd performance, reliability and to remove the obvious data races.

### Primary Change Themes

The changes principally cover four main themes:

1) SSH handshaking (`client_listener.go`)

It was observed that SSH handshaking was overloading the CPU. By limiting the number of concurrent SSH handshakes, we've been able to make sure the `rportd` server has available resources for actually starting clients.

2) Server client and related packages locks (`server/clients/client.go`, `cr.go`)

Aspects of accessing and updating client state are performed by most of the `rportd` go-routines. Access to the data by these go-routines must therefore be thread-safe. Improving the locks when accessing client related data has removed many of the previous data races.

3) Client-side backoff and retry strategies  (`'clients/client.go`)

While clients did implement some backoff and retry, it was relatively basic. More conservative delays and much more jitter have been introduced. Additionally, the various timeouts, wait durations, etc have been moved to constants and may moved to config options in the future.

4) sqlite multi-threaded enhancements for the clients.db (`server/clients/sqlite.go`)

With only 1 max concurrent connection to the sqlite `clients.db` it was noticed that writes were getting queued and write times were multi-second. By allowing more than 1 max concurrent connection and retrying if sqlite returns busy, write times are back at millisecond timeframes.

### PR Approach

While it would have been nice to separate these changes into individual PRs, the nature of the discovery of main issues and subsequent emerging strategies for fixes means that this would have created a difficult source code management and commit activity. Therefore all the changes exist in this one PR. The strategies addressing each theme have been introduced to another senior developer via online code review sessions to accelerate the understanding of the nature and rationale for each change.

### Outcomes

It seems that the implemented changes seem to have improved the situation considerably. `rportd` will now start 15K clients without issue or data races. debug builds can handle 5K clients without problems. Clients seems to be connecting and disconnecting without issues and we are no longer seeing queues of sqlite database writes.

KonradKuznicki commented:
@morestatic It looks like Client setters are mostly used in test, maybe instead locking and unlocking in setters we could request a copy of the entire structure?

morestatic commented:
hi @KonradKuznicki, as per our follow up phone call, let's consider splitting the client into `client details` vs `client state`. the `client state` should use the getters and setters and the `client details` could use a copied `struct`.  but let's do this as part of a new PR with potentially some other concurrency improvements that we can discuss / agree with @thorstenkramm .

morestatic commented on client/client_test.go:
```diff
@@ -428,11 +428,11 @@ func (m *mockServer) IsConnected() bool {
 			return nil
 		}
-		time.Sleep(2 * time.Millisecond)
+		time.Sleep(200 * time.Millisecond)
```
i think given this is just a test that the hard-coded value is ok. let's also keep an eye on it. if causing any flaky tests then we might need something more tunable.

morestatic commented:
hi @KonradKuznicki , i've updated Thorsten on our conversation and added a note above. can you approve the PR and i'll do the merge in the morning?
ps. also sent this note via Teams but guessing you might not see it.
